Examples/Algorithms/Fatras/src/FatrasAlgorithm.cpp

```diff
@@ -0,0 +1,249 @@
+
+#include "ActsExamples/Fatras/FatrasAlgorithm.hpp"
+
+#include "Acts/MagneticField/SharedBField.hpp"
+#include "Acts/Propagator/EigenStepper.hpp"
+#include "Acts/Propagator/Navigator.hpp"
+#include "Acts/Propagator/Propagator.hpp"
+#include "Acts/Propagator/StraightLineStepper.hpp"
+#include "Acts/Surfaces/Surface.hpp"
+#include "ActsExamples/EventData/SimHit.hpp"
+#include "ActsExamples/EventData/SimParticle.hpp"
+#include "ActsExamples/Framework/BareAlgorithm.hpp"
+#include "ActsExamples/Framework/RandomNumbers.hpp"
+#include "ActsExamples/Framework/WhiteBoard.hpp"
+#include "ActsFatras/Kernel/PhysicsList.hpp"
+#include "ActsFatras/Kernel/PointLikePhysicsList.hpp"
+#include "ActsFatras/Kernel/Process.hpp"
+#include "ActsFatras/Kernel/Simulator.hpp"
+#include "ActsFatras/Physics/Decay/NoDecay.hpp"
+#include "ActsFatras/Physics/EnergyLoss/BetheBloch.hpp"
+#include "ActsFatras/Physics/EnergyLoss/BetheHeitler.hpp"
+#include "ActsFatras/Physics/Scattering/Highland.hpp"
+#include "ActsFatras/Physics/StandardPhysicsLists.hpp"
+#include "ActsFatras/Selectors/ChargeSelectors.hpp"
+#include "ActsFatras/Selectors/KinematicCasts.hpp"
+#include "ActsFatras/Selectors/SelectorHelpers.hpp"
+#include "ActsFatras/Selectors/SurfaceSelectors.hpp"
+
+namespace {
+
+/// Simple struct to select surfaces where hits should be generated.
+struct HitSurfaceSelector {
+  bool sensitive = false;
+  bool material = false;
+  bool passive = false;
+
+  /// Check if the surface should be used.
+  bool operator()(const Acts::Surface &surface) const {
+    // sensitive/material are not mutually exclusive
+    bool isSensitive = surface.associatedDetectorElement();
+    bool isMaterial = surface.surfaceMaterial();
+    // passive should be an orthogonal category
+    bool isPassive = not(isSensitive or isMaterial);
+    return (isSensitive and sensitive) or (isMaterial and material) or
+           (isPassive and passive);
+  }
+};
+
+}  // namespace
+
+// Same interface as `ActsFatras::Simulator` but with concrete types.
+struct ActsExamples::detail::FatrasAlgorithmSimulation {
+  virtual ~FatrasAlgorithmSimulation(){};
+  virtual Acts::Result<std::vector<ActsFatras::FailedParticle>> simulate(
+      const Acts::GeometryContext &, const Acts::MagneticFieldContext &,
+      ActsExamples::RandomEngine &, const ActsExamples::SimParticleContainer &,
+      ActsExamples::SimParticleContainer::sequence_type &,
+      ActsExamples::SimParticleContainer::sequence_type &,
+      ActsExamples::SimHitContainer::sequence_type &) const = 0;
+};
+
+namespace {
+
+// Magnetic-field specific PIMPL implementation.
+template <typename magnetic_field_t>
+struct FatrasAlgorithmSimulationT final
+    : ActsExamples::detail::FatrasAlgorithmSimulation {
+  using CutPMin = ActsFatras::Min<ActsFatras::Casts::P>;
+
+  // typedefs for charge particle simulation
+  // propagate charged particles numerically in the given magnetic field
+  using ChargedStepper = Acts::EigenStepper<magnetic_field_t>;
```

Comment:
I template on the magnetic field, because that is the only thing that is configurable. I would prefer not to add more configurability since it is not actually used. This will also change again once the magnetic field provider interface is merged.

Reply:
Since the same issue came up in the KF, please note down at least a comment that this is restricted to extrapolations in vacuum. This will save some time spent in debugging

Reply:
Comment added.